README.md

## What It Is

This software is a LiveWhale application module that allows other applications to subscribe to push notifications of updates of news, events and blurbs through a simple callback protocol. It is based on a light version of the [PubSubHubbub protocol](http://code.google.com/p/pubsubhubbub/) akin to that used by [Instagr.am](http://instagram.com/) for their [real-time updates](http://instagram.com/developer/realtime/).

### The Process

Web applications subscribe to updates of a certain type (news, events, blurbs) and an optional group id and/or tag. From that point forward, as content matching a LiveWhale Push subscription is created, updated or deleted, LiveWhale Push sends a notification to the callback_url provided at the time the subscription was made.

The notification is a JSON array of notification objects each including details about the subscription being matched as well as the LiveWhale item id, and if the item is new, was deleted, or what fields changed. It does not send the entire record as in some cases that would be unnecessary.

The application can then react to the notification, obtaining additional data through an external API if needed.

## Requirements

### LiveWhale

Although this software might work with earlier versions, it has only been tested with LiveWhale 1.4.1 or better. Version 0.0.3 of LiveWhale Push also requires the presence of LiveWhale Places.

### Database Tables

You must add two new tables to your LiveWhale database as follows:

#### API Client Table

    livewhale_apiclients

    id:             integer (11)
    created_at:     datetime
    updated_at:     datetime
    client_id:      varchar (32) latin1_general_ci
    client_secret:  varchar (32) latin1_general_ci
    name:           varchar (100) utf8_general_ci
    email:          varchar (100) latin1_general_ci

    index on:       client_id
    index_on:       client_id_client_secret


#### Subscriptions Table

    livewhale_hubsubscriptions

    id:             integer (11)
    created_at:     datetime
    client_id:      integer (11)
    callback_url:   varchar (255) latin1_general_ci
    object:         varchar (20) latin1_general_ci
    group_id:       integer (11), default: NULL
    place_id:       integer (11), default: NULL
    radius:         double unsigned, default: NULL
    tag:            varchar (100) utf8_general_ci, default: NULL

    index on:       object_tag
    index on:       object_group_id
    index_on:       object_group_id_tag
    index on:       object_place_id
    index on:       client_id

### Dependencies

You must install our [utilities classes](https://github.com/lewisandclark/utilities) to be able to use this module. It provides http status responses and  the inflector, which is used to handle pluralization of tags.

### API

This software only provides the means of updating web applications that an piece of content matching some parameters has changed. As it cannot fathom what your web application might do with this knowledge, it does not send the data itself. There is one built in option for retrieving key fields that are normally available through widgets (outlined in Notifications > Content Retrieval below).

However, if you need a greater level of data, or access to data that is not currently live (i.e. hidden or scheduled), there isn't an official LiveWhale API yet. For our use, we have simply grafted a Rails app onto the LiveWhale tables as a read-only REST interface. (LiveWhale, the Rails app and our database are all on the same server, but it need not be that way.) I'd recommend this path as a temporary solution if you know Rails, as it only takes a little while to setup.

## Installation

The easiest way to install this software is to use git to clone it into your livewhale/client/modules folder as follows:

    $ cd /path/to/your/livewhale/client/modules
    $ git clone git://github.com/lewisandclark/push.git

Git will then copy the most current version of the code into a push folder within client/modules.

If you don't have or are unable to use git, you can also download a zip or tarball from github (use the downloads button) and extract it manually into the livewhale/client/modules folder as push. (Don't change the name, it will make it non-functional in LiveWhale.)

## API Clients

You will need to create API clients by hand in your database as a helper tool is not yet available (planned for a future release). To do this, insert into your livewhale_apiclients a row containing two different 32 character hashes for both the `client_id` and `client_secret` and the obvious other elements.

If you are using MySQL, the following code would work (replace `client_id`, `client_secret`, `name` and `email` with your values):

    INSERT INTO `livewhale_apiclients` (`id`, `created_at`, `updated_at`, `client_id`, `client_secret`, `name`, `email`) VALUES(NULL, NOW(), NOW(), '0123456798abcdef0123456798abcdef', '0123456798abcdef0123456798abcdef', 'Some Application Name', 'contact.or.errors@some.application.com');

## Usage

Once you have everything installed and have created your first API client, you will then be able to use the following usage guide to manage LiveWhale Push notifications.

### Subscription

Web applications may use the three `/live/` REST urls to be able to manage their content subscriptions.

#### Subscribe

To create a subscription, you will need to POST (at minimum) your `client_id`, `client_secret`, `object` and `callback_url` to the `/live/subscribe` url. (Https is recommended to maintain the security of your `client_secret`, but this is not enforced.) You may additionally add a `group_id`, a `place_id` and `radius`, a `tag` and a `verify_token`.

Please note:

  * Tag requests are automatically canonicalized; tags are reduced to the singular form or the word, all lower case.
  * Specifying a `place_id` without a `radius` assumes a zero radius, i.e. that place only.
  * Remember to encode your callback_url if non-basic-latin characters are present.
  * The `verify_token` can be used by you to differentiate similar subscriptions. It will be returned with the callback test only. It is not stored.

If your application does not need active management of subscriptions (i.e. they don't change often) then you can simply use cURL to manage them from the command line. An example follows:

    curl \
      -F 'client_id=0123456798abcdef0123456798abcdef' \
      -F 'client_secret=0123456798abcdef0123456798abcdef' \
      -F 'object=news' \
      -F 'tag=audio' \
      -F 'group_id=3' \
      -F 'place_id=29' \
      -F 'radius=0.75' \
      -F 'verify_token=some-verification-token' \
      -F 'callback_url=http://some.application.com/subscription/callback' \
      https://your.livewhale.com/live/subscribe/

LiveWhale Push will call your callback as part of the subscription process to test it. This will be a GET request with a `hub_challenge` parameter. (Actual notifications arrive as POSTs, see below.) At minimum, all your web application needs to do to verify the callback for LiveWhale Push is return the `hub_challenge` in the body of the response.

If you wish, you can authenticate the request using the `verify_token` or handle additional issues before returning the `hub_challenge`. If you need or want the highest level of security, you can confirm that the `hub_challenge` matches the SHA1 digest of your initial subscription request ip address and your `callback_url` combined in that order with two dashes between, and encrypted with your `client_secret` as the secret.

If your subscription request is successfully created, you will receive a JSON array with a single subscription object. The subscription object (shown below) contains the subscription `id` and your `client_id`, as well as the `object`, `group_id` and `tag` for the subscription (regardless of whether they were specified or not). Otherwise, you will receive a http status code error along with an error message as to the problem encountered.

    [
      {
        id: 9,
        client_id: '0123456798abcdef0123456798abcdef',
        object: 'news',
        group_id: 3,
        place_id: 29,
        radius: 0.75,
        tag: 'audio'
      }
    ]

Excluding `verify_token`, if you attempt to create another subscription with the same parameters, LiveWhale Push will merely return the original subscription JSON as above. Any change in any parameter will then create a new subscription. (In other words, subscriptions cannot be updated; they can only be created and deleted.)

#### Unsubscribe

To unsubscribe from LiveWhale Push, submit a POST request to `/live/unsubscribe` with your `client_id`, `client_secret` and `id`, where `id` is the subscription id you wish to terminate. (Https is recommended for security, but not required.) Again, unless your application requires active management of the subscriptions, you can use cURL to handle this:

    curl \
      -F 'client_id=0123456798abcdef0123456798abcdef' \
      -F 'client_secret=0123456798abcdef0123456798abcdef' \
      -F 'id=10' \
      https://your.livewhale.com/live/unsubscribe/

If successful, you will receive a http status code of 200. Otherwise, you will receive a http status code error along with an error message as to the problem encountered.

#### List Subscriptions

To see your current subscriptions, submit a POST request to `/live/subscriptions` with your `client_id` and `client_secret`. (Https is recommended for security, but not required.) Again, unless your application requires active management of the subscriptions, you can use cURL to handle this:

    curl \
      -F 'client_id=0123456798abcdef0123456798abcdef' \
      -F 'client_secret=0123456798abcdef0123456798abcdef' \
      https://your.livewhale.com/live/subscriptions/

If successful, you will receive a JSON array of subscription objects or an empty array if no subscriptions exist. Otherwise, you will receive a http status code error along with an error message as to the problem encountered.

### Notifications

Once a subscription has been created, LiveWhale Push monitors all CRUD actions (creates, updates and deletes) to catch those that match your `object`, `group_id` and `tag` parameters. When a match is found, it will POST a JSON array of notification updates to your callback_url. Your web application should be prepared to receive the following (example shown):

    [
      {
        subscription_id: 9,
        object: 'news',
        object_id: 7043
        group_id: 3,
        place_id: 29,
        radius: 0.75,
        tag: 'audio',
        updated_at: '2011-06-14T14:53:27-07:00',
        is_new: false,
        is_deleted: false,
        changed: ['search_tags']
      },...
    ]

You may receive one or more updates at a time, so your application should be prepared for that possibility. Also, in the event that an item has a watched `tag` added or subtracted -- thus entering or exiting a watched status -- you will receive a notification in all cases so that your application can sort out an appropriate action.

In the event that LiveWhale Push cannot access your callback for any reason, it will send an email your the API client's email address with an error message along with any data that was being pushed, in whatever state it might be at the time of the error.

#### Content Retrieval 

If the content is live (i.e. not hidden or scheduled), you can retrieve the widget-available fields through the `/live/` request methods. For example, if you received an update for event id #6117, you could get that content as JSON or XML:

    http://your.livewhale.com/live/events/6114@JSON
    http://your.livewhale.com/live/events/6114@XML

The JSON version of the above request produces (from our LiveWhale instance):

    {
      id: '6114',
      title: 'Oregon Bus Project',
      date: '06/13/2011',
      date_time: null,
      date2: null,
      date2_time: null,
      repeats: null,
      repeats_until: null,
      summary: '<a href="http://busproject.org/">Summer Conference</a>',
      description: '<p>\n  <a href="http://busproject.org/">Summer Conference</a>\n</p>',
      location: 'JRHH 202 &amp; Grape Arbor',
      date_created: '2011-06-10 15:32:31',
      last_modified: '2011-06-14 09:39:33',
      has_registration: null,
      image: null,
      url: http://www.lclark.edu/live/events/6114-oregon-bus-project'
    }

The XML version produces the same result in XML with null elements dropped:

    <result>
      <id>6114</id>
      <title>Oregon Bus Project</title>
      <date>06/13/2011</date>
      <summary>&lt;a href="http://busproject.org/"&gt;Summer Conference&lt;/a&gt;</summary>
      <description>&lt;p&gt;
      &lt;a href="http://busproject.org/"&gt;Summer Conference&lt;/a&gt;
    &lt;/p&gt;</description>
      <location>JRHH 202 &amp; Grape Arbor</location>
      <date_created>2011-06-10 15:32:31</date_created>
      <last_modified>2011-06-14 09:39:33</last_modified>
      <url>http://www.lclark.edu/live/events/6114-oregon-bus-project</url>
    </result>

As noted above, if you need a greater level of access than this, you'll need an API or other method to retrieve data.

## Developers

To my knowledge, this is the first LiveWhale application module available as open source. If you have suggestions, contributions, errors, etc. please email me, or register an issue, or fork from the DEV branch and issue a pull request. If you add additional functionality, your pull request must have corresponding supporting documentation.


HISTORY.md

0.0.3 / 2011-08-05 
==================

  * Added support for LiveWhale Places
  * Fixed naming bugs in /live/ REST api responders

0.0.1 / 2011-06-14 
==================

  * Initial release

private.application.push.php
<?php

$_LW->REGISTERED_APPS['push']=array(
	'title'=>'Push', // the module name
	'handlers' => array('init','save_success','manager_submit','delete')
);

// ----- PASS-THRU HANDLERS ---------------------------------------------------------

class LiveWhaleApplicationPush {

  public function init () {
    global $_LW;
  	include_once($_LW->INCLUDES_DIR_PATH . '/client/modules/push/includes/class.livewhalepush.php');
    $_LW->REGISTERED_APPS['push']['object'] = new LiveWhalePush();
    return NULL;
  }
  
  public function saveSuccess () {
    global $_LW;
    return $_LW->REGISTERED_APPS['push']['object']->save_success();
  }
  
  public function managerSubmit () {
    global $_LW;
    return $_LW->REGISTERED_APPS['push']['object']->manager_submit();
  }
  
  public function delete () {
    global $_LW;
    return $_LW->REGISTERED_APPS['push']['object']->delete();
  }

}

?>

includes/class.push.php
<?php

require_once($_LW->INCLUDES_DIR_PATH.'/client/utilities/class.inflector.php');
require_once($_LW->INCLUDES_DIR_PATH.'/client/utilities/class.httpstatuscodes.php');

class PubSubHub {

  /* database tables */
  static $_subscription_table = 'livewhale_hubsubscriptions';
  static $_subscription_columns = array(
    'id',
    'created_at',
    'client_id',
    'object',
    'callback_url'
    );
  static $_clients_table = 'livewhale_apiclients';

  /* api settings */
  static $_api_client_id_length = 32;
  static $_api_client_secret_length = 32;

  /* pubsub settings */
  static $_subscription_objects = array(
    'news',
    'events',
    'blurbs'
    );
  static $_subscription_aspects = array(
    'tag' => array(
      'allow' => 'alphanumeric characters plus dash or underscore',
      'transform' => "return preg_replace('~[^a-z\d_\-]+~i', '', \$value);"
      ),
    'group_id' => array(
      'allow' => 'an integer',
      'transform' => "return (int) \$value;"
      )
    );
  static $_subscription_public_columns = array(
    'id',
    'object',
    'group_id',
    'tag',
    'callback_url'
    );

  /* curl settings */
	protected $_curl_defaults = array(
		CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
		CURLOPT_CONNECTTIMEOUT_MS => 5000,
		CURLOPT_TIMEOUT_MS => 10000,
		CURLOPT_SSL_VERIFYPEER => false,
		CURLOPT_SSL_VERIFYHOST => false,
		CURLOPT_HEADER => true,
		CURLOPT_RETURNTRANSFER => true
		);

  protected $_watching = FALSE;
  protected $_object_type = "";
  protected $_query = "";
  protected $_before_update = array();
  protected $_delete_ids = array();
  protected $_subscriptions = array();

  /* initialization */

  public function __construct () {
  	global $_LW;
    $this->_object_type = preg_replace('~_(edit|list)$~', '', $_LW->page);
    if ( in_array($this->_object_type, PubSubHub::$_subscription_objects) ) $this->_watching = TRUE;
    if ( $this->_watching ) {
      if ( !empty($_GET['d']) ) { // editor delete
        $delete_id = preg_replace('~[^\d]+~', '', $_GET['d']);
        if ( $_GET['d'] == $delete_id ) {
          $this->_delete_ids = array($delete_id);
          $this->_query = "SELECT * FROM `livewhale_{$this->_object_type}` WHERE `id` = {$delete_id};";
        }
      } else if ( !empty($_LW->_POST['dropdown_checked']) && !empty($_LW->_POST['items']) ) { // manager dropdown action
        if ( $_LW->_POST['dropdown_checked'] == "{$this->_object_type}_delete" ) $this->_delete_ids = $_LW->_POST['items'];
        $this->_query = "SELECT * FROM `livewhale_{$this->_object_type}` WHERE `id` IN (" . implode(',', $_LW->_POST['items']) . ");";
      } else if ( !empty($_LW->_POST['item_id']) ) { // editor update
        $this->_query = "SELECT * FROM `livewhale_{$this->_object_type}` WHERE `id` = " . $_LW->_POST['item_id'] . ";";
      } else if ( !empty($_LW->_POST[$_LW->page]) && is_array($_LW->_POST[$_LW->page]) ) { // manager multi-click action
        $this->_query = "SELECT * FROM `livewhale_{$this->_object_type}` WHERE `id` IN (" . implode(',', $_LW->_POST[$_LW->page]) . ");";
      } // else create
    }
    if ( !empty($this->_query) ) $result = $_LW->query($this->_query);
    if ( !empty($result) && $result->num_rows ) while ( $item = $result->fetch_assoc() ) $this->_before_update[$item['id']] = $item;
  }

  /* Subscription Handling */
  private function require_valid_api_client ( $args ) {
    global $_LW;
    if ( array_key_exists('client_id', $args) ) $id = preg_replace('~[^a-f\d]~', '', $args['client_id']);
    if ( array_key_exists('client_secret', $args) ) $secret = preg_replace('~[^a-f\d]~', '', $args['client_secret']);
    if ( empty($id) || strlen($id) != PubSubHub::$_api_client_id_length ) HTTPStatusCodes::unauthorized("This action requires a valid API client id.");
    if ( empty($secret) || strlen($secret) != PubSubHub::$_api_client_secret_length ) HTTPStatusCodes::unauthorized("This action requires a valid API client secret.");
    $result = $_LW->query("SELECT * FROM `" . PubSubHub::$_clients_table . "` WHERE `client_id` = '{$id}' AND `client_secret` = '{$secret}';");
    if ( !empty($result) && $result->num_rows == 1 ) return $result->fetch_assoc();
    HTTPStatusCodes::unauthorized("This action requires a valid API user; your client id and secret were not found.");
  }

  private function require_valid_subscription_request ( &$args ) {
    if ( array_key_exists('id', $args) ) {
      if ( empty($args['id']) || (int) $args['id'] != $args['id'] ) HTTPStatusCodes::bad_request("Subscription requests require a valid integer for id when id is provided.");
      $args['id'] = (int) $args['id'];
    } else {
      if ( !array_key_exists('object', $args) || empty($args['object']) ) HTTPStatusCodes::bad_request("Subscription requests require an object.");
      if ( !in_array($args['object'], PubSubHub::$_subscription_objects) ) HTTPStatusCodes::bad_request("Subscription requests require an object to be one of: " . Inflector::to_sentence(PubSubHub::$_subscription_objects, ', ', ' or ') . ".");
      if ( !array_key_exists('callback_url', $args) || empty($args['callback_url']) ) HTTPStatusCodes::bad_request("Subscription requests require a callback_url.");
      if ( preg_replace('~[^a-z\d:/\-_\.]~i', '', $args['callback_url']) != $args['callback_url'] ) HTTPStatusCodes::bad_request("Subscription requests require a callback_url containing only alphanumeric characters plus colon, period, slash, dash and underscore.");
      foreach ( PubSubHub::$_subscription_aspects as $aspect => $details ) {
        if ( array_key_exists($aspect, $args) ) {
          if ( empty($args[$aspect]) ) HTTPStatusCodes::bad_request("Subscription requests that possess an aspect such as {$aspect} must provide a value; don't send it otherwise.");
          $value = $args[$aspect];
          if ( eval($details['transform']) != $args[$aspect] ) HTTPStatusCodes::bad_request("Subscription requests that possess an aspect of {$aspect} may contain only {$details['allow']}.");
          if ( !is_numeric($args[$aspect]) ) $this->as_tag($args[$aspect]);
        }
      }
    }
  }

  private function subscription_query ( $args, $lead = 'SELECT * FROM' ) {
    $query = "{$lead} `" . PubSubHub::$_subscription_table . "` WHERE `client_id` = {$args['client_id']}";
    if ( array_key_exists('id', $args) && !empty($args['id']) ) {
      $query .= " AND `id` = {$args['id']}";
    } else {
      $query .= " AND `object` = '{$args['object']}'";
      foreach ( PubSubHub::$_subscription_aspects as $aspect => $details ) {
        if ( (!array_key_exists($aspect, $args) || empty($args[$aspect])) && $lead == 'SELECT * FROM' ) {
          $query .= " AND `{$aspect}` IS NULL";
        } else if ( array_key_exists($aspect, $args) && !empty($args[$aspect]) ) {
          $value = $args[$aspect];
          $value = eval($details['transform']);
          $query .= " AND `{$aspect}` = " . ((is_int($value)) ? $value : "'{$value}'");
        }
      }
    }
    return "{$query};";
  }

  private function require_subscription_not_exist ( $args ) {
    global $_LW;
    $query = $this->subscription_query($args);
    $result = $_LW->query($query);
    if ( !empty($result) && $result->num_rows == 1 ) HTTPStatusCodes::ok($this->subscription_to_json($result->fetch_assoc()));
  }

  private function require_successful_challenge ( $args ) {
    $challenge = hash_hmac('sha1', "{$_SERVER['REMOTE_ADDR']}--{$args['callback_url']}", $args['client_secret']);
		$session = curl_init("{$args['callback_url']}?hub.mode=subscription&hub.challenge={$challenge}" . ((!empty($args['verify_token'])) ? "&hub.verify_token={$args['verify_token']}" : ""));
		if ( !is_resource($session) ) HTTPStatusCodes::server_error("We were unable to initialize a request to {$args['callback_url']} at this time.");
		curl_setopt_array($session, $this->_curl_defaults);
		$response = curl_exec($session);
    curl_close($session);
    $result = $this->is_valid_response($response, $args['callback_url']);
    if ( is_string($result) ) HTTPStatusCodes::server_error($result);
		$body = trim(array_pop(explode("\r\n\r\n", $response)));
		if ( $body != $challenge ) HTTPStatusCodes::bad_request("We were unable to complete a request to {$args['callback_url']} at this time; the challenge failed.");
	}

  private function as_column ( &$value, $key ) {
    $value = "`{$value}`";
  }

  private function as_value ( &$value, $key, $args ) {
    if ( array_key_exists($key, $args) && !empty($args[$key]) ) {
      $value = $args[$key];
      if ( array_key_exists($key, PubSubHub::$_subscription_aspects) && array_key_exists('transform', PubSubHub::$_subscription_aspects[$key]) && !empty(PubSubHub::$_subscription_aspects[$key]['transform']) ) $value = eval(PubSubHub::$_subscription_aspects[$key]['transform']);
      if ( !is_int($args[$key]) && !is_numeric($args[$key]) ) $value = "'{$value}'";
    } else if ( substr($key, -3) == '_at' ) {
      $value = "NOW()";
    }
  }

  private function as_tag ( &$value, $key ) {
    $value = Inflector::singularize(mb_strtolower($value, 'UTF-8'));
  }

  private function create_subscription ( $args ) {
    global $_LW;
    $columns = PubSubHub::$_subscription_columns;
    $columns = array_merge(array_keys(PubSubHub::$_subscription_aspects), $columns);
    $values = array_fill_keys($columns, 'NULL');
    array_walk($values, array($this, 'as_value'), $args);
    array_walk($columns, array($this, 'as_column'));
    $result = $_LW->query("INSERT INTO `" . PubSubHub::$_subscription_table . "` (" . implode(', ', $columns) . ") VALUES (" . implode(', ', $values) . ");");
    if ( empty($result) ) HTTPStatusCodes::server_error("We were unable to create your subscription at this time.");
		$result = $_LW->query('SELECT LAST_INSERT_ID() AS id;');
		if ( empty($result) || $result->num_rows < 1 ) {
      $query = $this->subscription_query($args, "DELETE FROM");
      $_LW->query($query);
		  HTTPStatusCodes::server_error("We were unable to create your subscription at this time.");
		}
	  $row = $result->fetch_assoc();
    $values['id'] = (int) $row['id'];
    HTTPStatusCodes::created($this->subscription_to_json($values));
  }

  private function delete_subscription ( $args ) {
    global $_LW;
    $query = $this->subscription_query($args, "DELETE FROM");
    $result = $_LW->query($query);
    var_dump($query, $result);
    if ( empty($result) ) HTTPStatusCodes::server_error("We were unable to delete your subscription at this time.");
    HTTPStatusCodes::ok("Subscription deleted.");
  }

  /* public subscription controls */
  public function subscribe ( $args = array() ) {
    while ( is_array($args) && count($args) == 1 && is_array($args[0]) ) $args = $args[0];
    $args['client'] = $this->require_valid_api_client($args);
    $this->require_valid_subscription_request($args);
    $args['client_id'] = (int) $args['client']['id'];
    $this->require_subscription_not_exist($args);
    $this->require_successful_challenge($args);
    $this->create_subscription($args);
  }

  public function unsubscribe ( $args = array() ) {
    while ( is_array($args) && count($args) == 1 && is_array($args[0]) ) $args = $args[0];
    $args['client'] = $this->require_valid_api_client($args);
    $this->require_valid_subscription_request($args);
    $args['client_id'] = (int) $args['client']['id'];
    $this->delete_subscription($args);
  }

  public function subscriptions ( $args = array() ) {
    global $_LW;
    while ( is_array($args) && count($args) == 1 && is_array($args[0]) ) $args = $args[0];
    $args['client'] = $this->require_valid_api_client($args);
    $args['client_id'] = (int) $args['client']['id'];
		$result = $_LW->query("SELECT * FROM `" . PubSubHub::$_subscription_table . "` WHERE `client_id` = {$args['client_id']};");
		if ( empty($result) || $result->num_rows < 1 ) HTTPStatusCodes::not_found("There were no subscriptions found.");
		$subscriptions = array();
		while ( $row = $result->fetch_assoc() ) $subscriptions[] = $this->public_subscription_for($row);
		HTTPStatusCodes::ok($this->to_json($subscriptions));
  }

  /* output methods */
  private function public_subscription_for ( $subscription = array() ) {
    $output = array();
    foreach ( PubSubHub::$_subscription_public_columns as $key ) {
      if ( array_key_exists($key, $subscription) && !empty($subscription[$key]) && $subscription[$key] != 'NULL' ) {
        $output[$key] = preg_replace('~^\'(.*)\'$~', '$1', $subscription[$key]);
      }
    }
    return $output;
  }

  private function to_json( $content = array() ) {
    if ( !empty($content) ) return json_encode($content);
    return "";
  }

  private function subscription_to_json ( $subscription = array() ) {
    return $this->to_json($this->public_subscription_for($subscription));
  }

  /* cURL response handling */

  private function is_valid_response ( $response, $callback_url ) {
		if ( $response === FALSE ) return "We were unable to complete a request to {$callback_url} at this time.";
		if ( !preg_match('~HTTP/1\.[01] ([\d]{3}) ([^\r\n]+)~', $response, $matches) ) return "We were unable to complete a request to {$callback_url} at this time; the server response could not be parsed.";
		if ( (int) $matches[1] != 200 ) return "We were unable to complete a request to {$callback_url} at this time; the server responded with: {$matches[1]} {$matches[2]}.";
		return true;
  }

  /* update handling */

  private function changed ( $before, $after ) {
    $changed = array();
    foreach ( $before as $key => $value ) if ( $key != 'last_modified' && $value != $after[$key] ) $changed[$key] = $after[$key];
    return $changed;
  }

  private function is_changed ( $before, $after ) {
    foreach ( $before as $key => $value ) if ( $key != 'last_modified' && $value != $after[$key] ) return TRUE;
    return FALSE;
  }

  private function find_subscriptions_for ( $object, $tags, $changed ) {
  	global $_LW;
    $query = "SELECT `" . PubSubHub::$_subscription_table . "`.*, `" . PubSubHub::$_clients_table . "`.`client_secret`, `" . PubSubHub::$_clients_table . "`.`email` FROM `" . PubSubHub::$_subscription_table . "` JOIN `" . PubSubHub::$_clients_table . "` ON `" . PubSubHub::$_subscription_table . "`.`client_id` = `" . PubSubHub::$_clients_table . "`.`id` WHERE `" . PubSubHub::$_subscription_table . "`.`object` = '{$this->_object_type}' AND (`" . PubSubHub::$_subscription_table . "`.`group_id` IS NULL" . ((!empty($object['gid'])) ? " OR `" . PubSubHub::$_subscription_table . "`.`group_id` = {$object['gid']}" : "") . ") AND (`" . PubSubHub::$_subscription_table . "`.`tag` is NULL" . ((!empty($tags)) ? " OR `" . PubSubHub::$_subscription_table . "`.`tag` = '" . implode("' OR `" . PubSubHub::$_subscription_table . "`.`tag` = '", $tags) . "'" : "") . ");";
		$result = $_LW->query($query);
  	if ( !empty($result) && $result->num_rows ) {
      while ( $subscription = $result->fetch_assoc() ) {
        $client_key = "{$subscription['client_secret']},{$subscription['email']}";
        if ( empty($this->_subscriptions[$client_key]) ) $this->_subscriptions[$client_key] = array();
        if ( empty($this->_subscriptions[$client_key]["{$subscription['callback_url']}"]) ) $this->_subscriptions[$client_key]["{$subscription['callback_url']}"] = array();
        $this->_subscriptions[$client_key]["{$subscription['callback_url']}"][] = array(
          'subscription_id' => (int) $subscription['id'],
          'object' => $this->_object_type,
          'object_id' => (int) $object['id'],
          'group_id' => ((empty($subscription['group_id'])) ? '' : (int) $subscription['group_id']),
          'tag' => ((empty($subscription['tag'])) ? '' : $subscription['tag']),
          'updated_at' => date("c"),
          'is_new' => (($changed['is_new']) ? TRUE : FALSE),
          'is_deleted' => (($changed['is_deleted']) ? TRUE : FALSE),
          'changed' => ((!$changed['is_new'] && !$changed['is_deleted']) ? array_keys($changed) : array())
        );
      }
    }
    return NULL;
  }

  private function notify_subscribers () {
    if ( empty($this->_subscriptions) ) return NULL;
    foreach ( (array) $this->_subscriptions as $client_key => $notifications ) {
      list($client_secret, $client_email) = explode(',', $client_key);
      foreach ( $notifications as $callback_url => $updates ) {
        $message = "";
        try {
          $json = json_encode($updates);
          $signature = hash_hmac('sha1', $json, $client_secret);
    		  $session = curl_init($callback_url);
    		  curl_setopt_array($session, $this->_curl_defaults);
    		  curl_setopt($session, CURLOPT_HTTPHEADER, array("X-Hub-Signature: {$signature}"));
    		  curl_setopt($session, CURLOPT_POSTFIELDS, "body={$json}");
    		  $response = curl_exec($session);
          curl_close($session);
          $result = $this->is_valid_response($response, $callback_url);
          if ( is_string($result) ) mail($client_email, 'LiveWhale PubSubHub Subscription Error', "{$result}\npayload:{$json}");
        } catch (Exception $e) {
          if ( !empty($json) ) {
      		  mail($client_email, 'LiveWhale PubSubHub Subscription Error', "Exception: {$e}\n\nThe data was: {$json}");
          } else {
      		  mail($client_email, 'LiveWhale PubSubHub Subscription Error', "Exception: {$e}\n\nThe data was: " . var_export($updates, TRUE));
          }
        }
      }
    }
    return NULL;
  }

  /* LiveWhale Hooks */

  public function save_success () {
  	global $_LW;
  	if ( $this->_watching && !empty($this->_before_update[$_LW->_POST['item_id']]) && !empty($this->_query) ) {
      $result = $_LW->query($this->_query);
      if ( $result && $result->num_rows ) {
        $after = $result->fetch_assoc();
        $changed = $this->changed($this->_before_update[$after['id']], $after);
        if ( empty($changed) ) return NULL;
      }
    } else if ( $this->_watching && !empty($_LW->new_id) ) {
      $this->_query = "SELECT * FROM `livewhale_{$this->_object_type}` WHERE `id` = {$_LW->new_id};";
      $result = $_LW->query($this->_query);
      if ( $result && $result->num_rows ) {
        $after = $result->fetch_assoc();
        $changed = array('is_new' => TRUE);
      }
    }
    if ( !empty($after) ) {
      $tags = array_unique(array_merge(explode(',', $this->_before_update[$after['id']]['search_tags']), explode(',', $after['search_tags'])));
      $this->find_subscriptions_for($after, $tags, $changed);
      return $this->notify_subscribers();
    }
    return NULL;
  }

  public function manager_submit () {
  	global $_LW;
    if ( $this->_watching && !empty($this->_before_update) && !empty($this->_query) ) {
      if ( !empty($this->_delete_ids) ) {
        foreach ( $this->_delete_ids as $id ) $this->find_subscriptions_for($this->_before_update[$id], explode(',', $this->_before_update[$id]['search_tags']), array('is_deleted' => TRUE));
      } else {
        $result = $_LW->query($this->_query);
        if ( !empty($result) && $result->num_rows ) {
          while ( $after = $result->fetch_assoc() ) {
            $changed = $this->changed($this->_before_update[$after['id']], $after);
            if ( !empty($changed) ) {
              $tags = array_unique(array_merge(explode(',', $this->_before_update[$after['id']]['search_tags']), explode(',', $after['search_tags'])));
              $this->find_subscriptions_for($after, $tags, $changed);
            }
          }
        }
      }
      return $this->notify_subscribers();
    }
    return NULL;
  }

  public function delete () {
  	if ( $this->_watching && !empty($_GET['d']) && !empty($this->_delete_ids) ) {
      $this->find_subscriptions_for($this->_before_update[$this->_delete_ids[0]], explode(',', $this->_before_update[$this->_delete_ids[0]]['search_tags']), array('is_deleted' => TRUE));
      return $this->notify_subscribers();
  	}
  	return NULL;
  }

}

/* curl -F 'client_id=5c79002e5ca04de49f9a2bafedd0acb0' -F 'client_secret=54a393596d314fcb985ca796ebf0923e' -F 'object=news' -F 'tag=greens' -F 'group_id=10' -F 'verify_token=wooo-hooo' -F 'callback_url=http://www.lclark.edu/tools/subscription_test.php' http://www.lclark.edu/live/subscribe/ */

/* curl -F 'client_id=5c79002e5ca04de49f9a2bafedd0acb0' -F 'client_secret=54a393596d314fcb985ca796ebf0923e' -F 'object=news' -F 'tag=greens' -F 'group_id=13' -F 'callback_url=http://www.lclark.edu/tools/subscription_test.php' http://www.lclark.edu/live/unsubscribe/ */

/* curl -F 'client_id=5c79002e5ca04de49f9a2bafedd0acb0' -F 'client_secret=54a393596d314fcb985ca796ebf0923e' -F 'object=news' -F 'tag=greens' -F 'callback_url=http://www.lclark.edu/tools/subscription_test.php' http://www.lclark.edu/live/unsubscribe/ */

/* curl -F 'client_id=5c79002e5ca04de49f9a2bafedd0acb0' -F 'client_secret=54a393596d314fcb985ca796ebf0923e' -F 'id=5' http://www.lclark.edu/live/unsubscribe/ */

/* curl -F 'client_id=5c79002e5ca04de49f9a2bafedd0acb0' -F 'client_secret=54a393596d314fcb985ca796ebf0923e' http://www.lclark.edu/live/subscriptions/ */

?>